dali/python/nvidia/dali/_multiproc/shared_queue.py

```diff
@@ -0,0 +1,233 @@
+
+from typing import List, Optional
+import threading
+from nvidia.dali._multiproc import shared_mem
+from nvidia.dali._multiproc.messages import Structure, ShmMessage
+from nvidia.dali._multiproc.shared_batch import _align_up as align_up
+
+
+class QueueMeta(Structure):
+    _fields = ("capacity", "i"), ("size", "i"), ("begining", "i"), ("is_closed", "i")
+
+
+class ShmQueue:
+
+    """
+    Simple fixed capacity shared memory queue of fixed size messages. The queue fails with assertion error
+    if written to when full, attempt to get from empty queue blocks until data is available or queue was
+    closed.
+    """
+
+    MSG_CLASS = ShmMessage
+    ALIGN_UP_MSG = 4
+    ALIGN_UP_BUFFER = 4096
+
+    def __init__(self, mp, capacity):
+        self.lock = mp.Lock()
+        self.cv_empty = mp.Condition(self.lock)
+        self.capacity = capacity
+        self.meta = QueueMeta(capacity, 0, 0, 0)
+        self.meta_capacity = align_up(self.meta.get_size(), self.ALIGN_UP_MSG)
+        dummy_msg = self.MSG_CLASS()
+        self.msg_capacity = align_up(dummy_msg.get_size(), self.ALIGN_UP_MSG)
+        self.shm_capacity = align_up(self.meta_capacity + capacity * self.msg_capacity, self.ALIGN_UP_BUFFER)
+        self.shm = shared_mem.SharedMem.allocate(self.shm_capacity)
+        self.is_closed = False
+        self.init_offsets()
+        self.write_meta()
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state['msgs_offsets'] = None
+        state['shm'] = None
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self.init_offsets()
+
+    def init_offsets(self):
+        self.msgs_offsets = [i * self.msg_capacity + self.meta_capacity for i in range(self.capacity)]
+
+    def set_shm(self, shm):
+        assert self.shm is None and shm.capacity >= self.shm_capacity
+        self.shm = shm
+
+    def read_meta(self):
+        self.meta.unpack_from(self.shm.buf, 0)
+
+    def write_meta(self):
+        self.meta.pack_into(self.shm.buf, 0)
+
+    def read_msg(self, i):
+        offset = self.msgs_offsets[i]
+        msg = self.MSG_CLASS()
+        msg.unpack_from(self.shm.buf, offset)
+        return msg
+
+    def write_msg(self, i, msg):
+        offset = self.msgs_offsets[i]
+        msg.pack_into(self.shm.buf, offset)
+
+    def seal(self):
+        self.shm.seal()
+
+    def close(self):
+        if self.is_closed:
+            return
+        with self.lock:
+            self.read_meta()
+            self.is_closed = True
+            if not self.meta.is_closed:
+                self.meta.is_closed = 1
+                self.write_meta()
+                # Notify only one waiting worker about closing of the queue, the woken up worker
+                # will notify the next one. Avoid notify_all at this point, due to possible deadlock if
+                # one of the notified workers exited abruptly when waiting on cv_empty without proper releasing
```

Comment:
So if we notifying on one worker, how we avoid the deadlock? Still there could be a one process that crashed and hasn't released the underlying semaphore.

Reply:
The main way to avoid this deadlock is notifying this way only under "normal" exit circumstances, i.e. when pool close method is called and if at the point of cleanup all workers are alive. Otherwise, tracker thread won't bother to call queue.close. This seems to be reliable, though one could imagine that workers drop dead just after the aliveness check and before main process calls the queue.close.
So notifying one worker (and properly releasing resources in worker processes if they are not immediately killed by the OS) serve as kind of additional safety measures.

Reply:
Understood.